File: sdks/typescript/packages/sdk/src/payload-topo.ts
/**
 * Resolve the create payload into an ordered list of operations.
 *
 * The old SDK derived ordering from a static FK schema introspected from
 * the database. With factories owning entity creation the SDK no longer
 * has — or needs — that schema. What it does have is the create payload
 * itself, and that already contains complete dependency information:
 *
 *   * Each entity that other entities depend on declares `_alias: "name"`.
 *   * Each entity that depends on another carries `{ _ref: "name" }`
 *     somewhere in its field tree (top-level FK, nested data blob, list
 *     element — anywhere).
 *
 * Walking the payload to collect alias → owner and owner → {refs} gives
 * us the exact dependency graph. Kahn's topo sort over that graph
 * produces the `up` order; the reverse is the `down` order.
 *
 * Cycles in the alias graph raise INVALID_BODY. The old SDK used to
 * break cycles by nullifying nullable FKs; with factories the host owns
 * FK semantics and the SDK does not modify rows, so we surface cycles
 * to the caller instead.
 */
import { Errors } from './errors'

export interface CreateOp {
  model: string
  fields: Record<string, unknown>
  tempId: string
}

export interface ResolvedTree {
  ops: CreateOp[]
  /** alias → temp id assigned to the entity declaring that alias. */
  aliases: Record<string, string>
  /** alias → model name, used by teardown to pick the right factory. */
  aliasOwnerModel: Record<string, string>
  /** alias → list of aliases the owner depends on (may include unknowns). */
  aliasDependencies: Record<string, string[]>
}

const RESERVED_KEYS = new Set(['_alias', '_ref'])

function collectRefs(value: unknown, out: string[]): void {
  if (value && typeof value === 'object' && !Array.isArray(value)) {
    const obj = value as Record<string, unknown>
    const ref = obj._ref
    if (typeof ref === 'string') {
      out.push(ref)
      return
    }
    for (const v of Object.values(obj)) collectRefs(v, out)
    return
  }
  if (Array.isArray(value)) {
    for (const v of value) collectRefs(v, out)
  }
}

function resolveRefs(value: unknown, aliasToTempId: Record<string, string>): unknown {
  if (value && typeof value === 'object' && !Array.isArray(value)) {
    const obj = value as Record<string, unknown>
    const ref = obj._ref
    if (typeof ref === 'string') {
      const real = aliasToTempId[ref]
      return real !== undefined ? real : value
    }
    const out: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(obj)) {
      out[k] = resolveRefs(v, aliasToTempId)
    }
    return out
  }
  if (Array.isArray(value)) {
    return value.map((v) => resolveRefs(v, aliasToTempId))
  }
  return value
}

/**
 * Topo-sort a create payload into an ordered list of `CreateOp`.
 *
 * `create` is the dashboard's nested map `{ model: [entity, ...] }`.
 * Each entity is an object; `_alias` (declared by dependency targets)
 * and `_ref` (declared by dependents, anywhere in the field tree) are
 * the only reserved keys.
 *
 * Throws INVALID_BODY if the payload references an alias that is never
 * declared, or if the alias graph contains a cycle.
 */
export function resolvePayloadTree(
  create: Record<string, unknown>,
): ResolvedTree {
  if (!create || typeof create !== 'object' || Array.isArray(create)) {
    throw Errors.invalidBody('`create` must be an object keyed by model name')
  }

  // First pass: assign temp ids and collect alias declarations.
  type RawEntry = {
    model: string
    tempId: string
    entity: Record<string, unknown>
    alias: string | null
  }

  const rawEntries: RawEntry[] = []
  let counter = 0
  const aliases: Record<string, string> = {}
  const aliasOwnerModel: Record<string, string> = {}

  for (const [model, entities] of Object.entries(create as Record<string, unknown>)) {
    if (!Array.isArray(entities)) {
      throw Errors.invalidBody(
        `\`create.${model}\` must be a list of entity objects, got ${typeof entities}`,
      )
    }
    for (const entity of entities) {
      if (!entity || typeof entity !== 'object' || Array.isArray(entity)) {
        throw Errors.invalidBody(
          `\`create.${model}\` entries must be objects, got ${Array.isArray(entity) ? 'array' : typeof entity}`,
        )
      }
      const tempId = `__temp_${model}_${counter++}`
      const obj = entity as Record<string, unknown>
      const aliasRaw = obj._alias
      let alias: string | null = null
      if (typeof aliasRaw === 'string') {
        if (aliases[aliasRaw] !== undefined) {
          throw Errors.invalidBody(`duplicate _alias "${aliasRaw}"`)
        }
        aliases[aliasRaw] = tempId
        aliasOwnerModel[aliasRaw] = model
        alias = aliasRaw
      } else if (aliasRaw !== undefined && aliasRaw !== null) {
        throw Errors.invalidBody('"_alias" must be a string')
      }
      rawEntries.push({ model, tempId, entity: obj, alias })
    }
  }

  // Second pass: collect each entry's dependency aliases (for the topo
  // graph) and strip reserved keys from its field dict.
  const depsByTempId: Record<string, string[]> = {}
  const fieldsByTempId: Record<string, Record<string, unknown>> = {}
  const modelByTempId: Record<string, string> = {}

  for (const { model, tempId, entity } of rawEntries) {
    const deps: string[] = []
    const cleaned: Record<string, unknown> = {}
    for (const [key, value] of Object.entries(entity)) {
      if (RESERVED_KEYS.has(key)) continue
      collectRefs(value, deps)
      // Rewrite every `{ _ref: alias }` to the alias's temp id. The
      // handler later swaps temp ids for the real factory-returned ids
      // during `up`.
      cleaned[key] = resolveRefs(value, aliases)
    }
    const unknowns = deps.filter((a) => aliases[a] === undefined)
    if (unknowns.length > 0) {
      const sorted = Array.from(new Set(unknowns)).sort()
      throw Errors.invalidBody(
        `\`create.${model}\` references unknown alias(es): ${sorted.join(', ')}`,
      )
    }
    depsByTempId[tempId] = deps
    fieldsByTempId[tempId] = cleaned
    modelByTempId[tempId] = model
  }

  // Build the temp_id graph and topo-sort.
  const inDegree: Record<string, number> = {}
  for (const { tempId } of rawEntries) inDegree[tempId] = 0
  const edges: Record<string, string[]> = {}
  for (const [tempId, deps] of Object.entries(depsByTempId)) {
    const seen = new Set<string>()
    for (const depAlias of deps) {
      const depTempId = aliases[depAlias]!
      if (depTempId === tempId || seen.has(depTempId)) continue
      seen.add(depTempId)
      ;(edges[depTempId] ??= []).push(tempId)
      inDegree[tempId] = (inDegree[tempId] ?? 0) + 1
    }
  }

  // Kahn's, preserving payload order as the stable tie-breaker.
  const payloadOrder: Record<string, number> = {}
  rawEntries.forEach((e, i) => {
    payloadOrder[e.tempId] = i
  })

  const ready = Object.keys(inDegree)
    .filter((t) => inDegree[t] === 0)
    .sort((a, b) => payloadOrder[a]! - payloadOrder[b]!)

  const sortedTempIds: string[] = []
  while (ready.length > 0) {
    const tid = ready.shift()!
    sortedTempIds.push(tid)
    for (const next of edges[tid] ?? []) {
      inDegree[next] = (inDegree[next] ?? 0) - 1
      if (inDegree[next] === 0) ready.push(next)
    }
    ready.sort((a, b) => payloadOrder[a]! - payloadOrder[b]!)
  }

  if (sortedTempIds.length !== rawEntries.length) {
    const cycle = Object.entries(inDegree)
      .filter(([, deg]) => deg > 0)
      .map(([tid]) => tid)
      .sort((a, b) => payloadOrder[a]! - payloadOrder[b]!)
    const cycleModels = cycle.map((t) => modelByTempId[t]).join(', ')
    throw Errors.invalidBody(`cycle detected in _alias/_ref graph: ${cycleModels}`)
  }

  // Build CreateOp list in topo order.
  const aliasDependencies: Record<string, string[]> = {}
  for (const [alias, tempId] of Object.entries(aliases)) {
    aliasDependencies[alias] = [...(depsByTempId[tempId] ?? [])]
  }

  const ops: CreateOp[] = sortedTempIds.map((tid) => ({
    model: modelByTempId[tid]!,
    fields: fieldsByTempId[tid]!,
    tempId: tid,
  }))

  return { ops, aliases, aliasOwnerModel, aliasDependencies }
}

/**
 * Order models for teardown.
 *
 * With `aliasDependencies` available (newer refs tokens carry it), we
 * run the same Kahn's topo sort over models — derived from aggregating
 * each alias's dependencies — and return the *reverse* topo so children
 * are torn down before parents.
 *
 * Without it (older refs tokens), fall back to reversing the insertion
 * order of `refs` keys, which is what the SDK always did for factory
 * teardown.
 */
export function computeTeardownOrder(
  refs: Record<string, unknown[]>,
  aliasDependencies?: Record<string, string[]>,
  aliasOwnerModel?: Record<string, string>,
): string[] {
  const models = Object.keys(refs)

  if (!aliasDependencies || !aliasOwnerModel || Object.keys(aliasDependencies).length === 0) {
    return [...models].reverse()
  }

  const modelDeps: Record<string, Set<string>> = {}
  for (const m of models) modelDeps[m] = new Set()

  for (const [alias, deps] of Object.entries(aliasDependencies)) {
    const owner = aliasOwnerModel[alias]
    if (!owner || !(owner in modelDeps)) continue
    for (const depAlias of deps) {
      const depModel = aliasOwnerModel[depAlias]
      if (!depModel || depModel === owner) continue
      if (depModel in modelDeps) {
        modelDeps[owner]!.add(depModel)
      }
    }
  }

  const inDegree: Record<string, number> = {}
  for (const m of models) inDegree[m] = 0
  const adj: Record<string, string[]> = {}
  for (const [owner, deps] of Object.entries(modelDeps)) {
    for (const depModel of deps) {
      ;(adj[depModel] ??= []).push(owner)
      inDegree[owner] = (inDegree[owner] ?? 0) + 1
    }
  }

  const payloadOrder: Record<string, number> = {}
  models.forEach((m, i) => {
    payloadOrder[m] = i
  })

  const ready = models
    .filter((m) => inDegree[m] === 0)
    .sort((a, b) => payloadOrder[a]! - payloadOrder[b]!)
  const upOrder: string[] = []
  while (ready.length > 0) {
    const m = ready.shift()!
    upOrder.push(m)
    for (const next of adj[m] ?? []) {
      inDegree[next] = (inDegree[next] ?? 0) - 1
      if (inDegree[next] === 0) ready.push(next)
    }
    ready.sort((a, b) => payloadOrder[a]! - payloadOrder[b]!)
  }

  if (upOrder.length !== models.length) {
    // Shouldn't happen — cycles are rejected at `up`. Fall back to
    // registration order to avoid losing data.
    return [...models].reverse()
  }

  return [...upOrder].reverse()
}

File: sdks/java/autonoma-sdk/src/main/java/ai/autonoma/sdk/PayloadTopo.java
package ai.autonoma.sdk;

import ai.autonoma.sdk.types.CreateOp;
import ai.autonoma.sdk.types.ResolvedTree;

import java.util.*;

/**
 * Resolve the create payload into an ordered list of operations.
 *
 * <p>Ordering comes from the payload's {@code _alias}/{@code _ref} graph.
 * Each entity that others depend on declares {@code _alias: "name"}.
 * Each entity that depends on another carries {@code {"_ref": "name"}}
 * somewhere in its field tree. Walking the payload to collect these gives
 * the exact dependency graph. Kahn's topo sort produces the up order;
 * the reverse is the down order.
 */
public final class PayloadTopo {

    private static final Set<String> RESERVED_KEYS = Set.of("_alias", "_ref");

    private PayloadTopo() {}

    /**
     * Topo-sort a create payload into an ordered list of CreateOp.
     *
     * @param create the dashboard's nested map {model: [entity, ...]}
     * @return a ResolvedTree with ops in topological order
     */
    @SuppressWarnings("unchecked")
    public static ResolvedTree resolvePayloadTree(Map<String, Object> create) {
        if (create == null) {
            throw AutonomaError.invalidBody("`create` must be an object keyed by model name");
        }

        // First pass: assign temp ids and collect alias declarations.
        List<RawEntry> rawEntries = new ArrayList<>();
        int counter = 0;
        Map<String, String> aliases = new LinkedHashMap<>();
        Map<String, String> aliasOwnerModel = new LinkedHashMap<>();

        for (Map.Entry<String, Object> modelEntry : create.entrySet()) {
            String model = modelEntry.getKey();
            Object entitiesRaw = modelEntry.getValue();
            if (!(entitiesRaw instanceof List<?>)) {
                throw AutonomaError.invalidBody(
                    "`create." + model + "` must be a list of entity objects, got " + entitiesRaw.getClass().getSimpleName());
            }
            List<?> entities = (List<?>) entitiesRaw;
            for (Object entityRaw : entities) {
                if (!(entityRaw instanceof Map<?, ?>)) {
                    throw AutonomaError.invalidBody(
                        "`create." + model + "` entries must be objects, got " + entityRaw.getClass().getSimpleName());
                }
                Map<String, Object> entity = (Map<String, Object>) entityRaw;
                String tempId = "__temp_" + model + "_" + counter;
                counter++;

                Object aliasRaw = entity.get("_alias");
                String alias = null;
                if (aliasRaw instanceof String a) {
                    if (aliases.containsKey(a)) {
                        throw AutonomaError.invalidBody("duplicate _alias \"" + a + "\"");
                    }
                    aliases.put(a, tempId);
                    aliasOwnerModel.put(a, model);
                    alias = a;
                } else if (aliasRaw != null) {
                    throw AutonomaError.invalidBody("\"_alias\" must be a string");
                }

                rawEntries.add(new RawEntry(model, tempId, entity, alias));
            }
        }

        // Second pass: collect each entry's dependency aliases and strip reserved keys.
        Map<String, List<String>> depsByTempId = new LinkedHashMap<>();
        Map<String, Map<String, Object>> fieldsByTempId = new LinkedHashMap<>();
        Map<String, String> modelByTempId = new LinkedHashMap<>();
        Map<String, String> aliasByTempId = new LinkedHashMap<>();

        for (RawEntry entry : rawEntries) {
            List<String> deps = new ArrayList<>();
            Map<String, Object> cleaned = new LinkedHashMap<>();
            for (Map.Entry<String, Object> kv : entry.entity.entrySet()) {
                if (RESERVED_KEYS.contains(kv.getKey())) continue;
                collectRefs(kv.getValue(), deps);
                cleaned.put(kv.getKey(), resolveRefs(kv.getValue(), aliases));
            }
            List<String> unknown = new ArrayList<>();
            for (String a : deps) {
                if (!aliases.containsKey(a)) unknown.add(a);
            }
            if (!unknown.isEmpty()) {
                Set<String> uniqueUnknown = new TreeSet<>(unknown);
                throw AutonomaError.invalidBody(
                    "`create." + entry.model + "` references unknown alias(es): " + String.join(", ", uniqueUnknown));
            }
            depsByTempId.put(entry.tempId, deps);
            fieldsByTempId.put(entry.tempId, cleaned);
            modelByTempId.put(entry.tempId, entry.model);
            aliasByTempId.put(entry.tempId, entry.alias);
        }

        // Build the temp_id graph and topo-sort.
        Map<String, Integer> inDegree = new LinkedHashMap<>();
        for (RawEntry entry : rawEntries) {
            inDegree.put(entry.tempId, 0);
        }
        Map<String, List<String>> edges = new LinkedHashMap<>();

        for (Map.Entry<String, List<String>> entry : depsByTempId.entrySet()) {
            String tempId = entry.getKey();
            Set<String> seen = new HashSet<>();
            for (String depAlias : entry.getValue()) {
                String depTempId = aliases.get(depAlias);
                if (depTempId.equals(tempId) || seen.contains(depTempId)) continue;
                seen.add(depTempId);
                edges.computeIfAbsent(depTempId, k -> new ArrayList<>()).add(tempId);
                inDegree.merge(tempId, 1, Integer::sum);
            }
        }

        // Kahn's, preserving payload order as stable tie-breaker.
        Map<String, Integer> payloadOrder = new LinkedHashMap<>();
        for (int i = 0; i < rawEntries.size(); i++) {
            payloadOrder.put(rawEntries.get(i).tempId, i);
        }

        List<String> ready = new ArrayList<>();
        for (Map.Entry<String, Integer> e : inDegree.entrySet()) {
            if (e.getValue() == 0) ready.add(e.getKey());
        }
        ready.sort(Comparator.comparingInt(t -> payloadOrder.getOrDefault(t, 0)));

        List<String> sortedTempIds = new ArrayList<>();
        while (!ready.isEmpty()) {
            String tid = ready.remove(0);
            sortedTempIds.add(tid);
            for (String nxt : edges.getOrDefault(tid, List.of())) {
                int newDeg = inDegree.get(nxt) - 1;
                inDegree.put(nxt, newDeg);
                if (newDeg == 0) ready.add(nxt);
            }
            ready.sort(Comparator.comparingInt(t -> payloadOrder.getOrDefault(t, 0)));
        }

        if (sortedTempIds.size() != rawEntries.size()) {
            List<String> cycle = new ArrayList<>();
            for (Map.Entry<String, Integer> e : inDegree.entrySet()) {
                if (e.getValue() > 0) cycle.add(e.getKey());
            }
            cycle.sort(Comparator.comparingInt(t -> payloadOrder.getOrDefault(t, 0)));
            StringJoiner cycleModels = new StringJoiner(", ");
            for (String t : cycle) cycleModels.add(modelByTempId.get(t));
            throw AutonomaError.invalidBody("cycle detected in _alias/_ref graph: " + cycleModels);
        }

        // Build CreateOp list and alias dependencies.
        Map<String, List<String>> aliasDependencies = new LinkedHashMap<>();
        for (String alias : aliases.keySet()) {
            String tid = aliases.get(alias);
            aliasDependencies.put(alias, new ArrayList<>(depsByTempId.getOrDefault(tid, List.of())));
        }

        List<CreateOp> ops = new ArrayList<>();
        for (String tid : sortedTempIds) {
            ops.add(new CreateOp(modelByTempId.get(tid), fieldsByTempId.get(tid), tid));
        }

        return new ResolvedTree(ops, aliases, aliasOwnerModel, aliasDependencies);
    }

    /**
     * Compute model teardown order from alias dependency info in the refs token.
     */
    @SuppressWarnings("unchecked")
    public static List<String> computeTeardownOrder(
            Map<String, List<Map<String, Object>>> refs,
            Map<String, Object> aliasDependenciesRaw,
            Map<String, Object> aliasOwnerModelRaw) {

        List<String> models = new ArrayList<>(refs.keySet());

        if (aliasDependenciesRaw == null || aliasDependenciesRaw.isEmpty()
                || aliasOwnerModelRaw == null || aliasOwnerModelRaw.isEmpty()) {
            List<String> reversed = new ArrayList<>(models);
            Collections.reverse(reversed);
            return reversed;
        }

        // Convert raw maps to typed maps
        Map<String, List<String>> aliasDependencies = new LinkedHashMap<>();
        for (Map.Entry<String, Object> e : aliasDependenciesRaw.entrySet()) {
            List<String> deps = new ArrayList<>();
            if (e.getValue() instanceof List<?> list) {
                for (Object item : list) deps.add(item.toString());
            }
            aliasDependencies.put(e.getKey(), deps);
        }
        Map<String, String> aliasOwnerModel = new LinkedHashMap<>();
        for (Map.Entry<String, Object> e : aliasOwnerModelRaw.entrySet()) {
            aliasOwnerModel.put(e.getKey(), e.getValue().toString());
        }

        // Build model -> {model dependencies}
        Map<String, Set<String>> modelDeps = new LinkedHashMap<>();
        for (String m : models) modelDeps.put(m, new LinkedHashSet<>());

        for (Map.Entry<String, List<String>> entry : aliasDependencies.entrySet()) {
            String alias = entry.getKey();
            String owner = aliasOwnerModel.get(alias);
            if (owner == null || !modelDeps.containsKey(owner)) continue;
            for (String depAlias : entry.getValue()) {
                String depModel = aliasOwnerModel.get(depAlias);
                if (depModel == null || depModel.equals(owner)) continue;
                if (modelDeps.containsKey(depModel)) {
                    modelDeps.get(owner).add(depModel);
                }
            }
        }

        // Kahn's over models.
        Map<String, Integer> inDegree = new LinkedHashMap<>();
        Map<String, List<String>> adj = new LinkedHashMap<>();
        for (String m : models) {
            inDegree.put(m, 0);
            adj.put(m, new ArrayList<>());
        }
        for (Map.Entry<String, Set<String>> entry : modelDeps.entrySet()) {
            String owner = entry.getKey();
            for (String depModel : entry.getValue()) {
                adj.get(depModel).add(owner);
                inDegree.merge(owner, 1, Integer::sum);
            }
        }

        Map<String, Integer> payloadOrder = new LinkedHashMap<>();
        for (int i = 0; i < models.size(); i++) payloadOrder.put(models.get(i), i);

        List<String> ready = new ArrayList<>();
        for (Map.Entry<String, Integer> e : inDegree.entrySet()) {
            if (e.getValue() == 0) ready.add(e.getKey());
        }
        ready.sort(Comparator.comparingInt(m -> payloadOrder.getOrDefault(m, 0)));

        List<String> upOrder = new ArrayList<>();
        while (!ready.isEmpty()) {
            String m = ready.remove(0);
            upOrder.add(m);
            for (String nxt : adj.getOrDefault(m, List.of())) {
                int newDeg = inDegree.get(nxt) - 1;
                inDegree.put(nxt, newDeg);
                if (newDeg == 0) ready.add(nxt);
            }
            ready.sort(Comparator.comparingInt(mm -> payloadOrder.getOrDefault(mm, 0)));
        }

        if (upOrder.size() != models.size()) {
            // Shouldn't happen - cycles were rejected at up. Fall back.
            List<String> reversed = new ArrayList<>(models);
            Collections.reverse(reversed);
            return reversed;
        }

        List<String> result = new ArrayList<>(upOrder);
        Collections.reverse(result);
        return result;
    }

    // --- internal helpers ---

    @SuppressWarnings("unchecked")
    private static void collectRefs(Object value, List<String> out) {
        if (value instanceof Map<?, ?> map) {
            Object ref = map.get("_ref");
            if (ref instanceof String s) {
                out.add(s);
                return;
            }
            for (Object v : map.values()) collectRefs(v, out);
        } else if (value instanceof List<?> list) {
            for (Object v : list) collectRefs(v, out);
        }
    }

    @SuppressWarnings("unchecked")
    private static Object resolveRefs(Object value, Map<String, String> aliasToTempId) {
        if (value instanceof Map<?, ?> map) {
            Object ref = map.get("_ref");
            if (ref instanceof String s) {
                String real = aliasToTempId.get(s);
                return real != null ? real : value;
            }
            Map<String, Object> result = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                result.put((String) e.getKey(), resolveRefs(e.getValue(), aliasToTempId));
            }
            return result;
        }
        if (value instanceof List<?> list) {
            List<Object> result = new ArrayList<>(list.size());
            for (Object v : list) result.add(resolveRefs(v, aliasToTempId));
            return result;
        }
        return value;
    }

    private record RawEntry(String model, String tempId, Map<String, Object> entity, String alias) {}
}
